src/Storages/RemoteReadBufferCache.h

```diff
@@ -0,0 +1,219 @@
+    bool checkFileChanged(IRemoteFileMetadataPtr file_metadata_);
+    inline void markInvalid()
+    {
+        std::lock_guard lock(mutex);
+        valid = false;
+    }
+    inline bool isValid()
+    {
+        std::lock_guard lock(mutex);
+        return valid;
+    }
+    IRemoteFileMetadataPtr getFileMetadata() { return file_metadata_ptr; }
+    inline size_t getFileSize() const { return file_metadata_ptr->getFileSize(); }
+
+    void startBackgroundDownload(std::unique_ptr<ReadBuffer> in_readbuffer_, BackgroundSchedulePool & thread_pool);
+
+private:
+    // flush file and status information
+    void flush(bool need_flush_status = false);
+    bool loadInnerInformation(const std::filesystem::path & file_path);
+
+    BackgroundSchedulePool::TaskHolder download_task_holder;
+    void backgroundDownload(ReadBufferPtr remote_read_buffer);
+
+    std::mutex mutex;
+    std::condition_variable more_data_signal;
+
+    std::set<uintptr_t> opened_file_buffer_refs; // refer to a buffer address
+
+    String metadata_class;
+    LocalFileStatus file_status = TO_DOWNLOAD; // for tracking download process
+    IRemoteFileMetadataPtr file_metadata_ptr;
+    std::filesystem::path local_path;
+
+    bool valid;
+    size_t local_cache_bytes_read_before_flush;
+    size_t current_offset;
+
+    //std::shared_ptr<ReadBuffer> remote_read_buffer;
+    std::unique_ptr<WriteBufferFromFileBase> data_file_writer;
+
+    Poco::Logger * log = &Poco::Logger::get("RemoteCacheController");
+};
+using RemoteCacheControllerPtr = std::shared_ptr<RemoteCacheController>;
+
+/*
+ * FIXME:RemoteReadBuffer derive from SeekableReadBufferWithSize may cause some risks, since it's not seekable in some cases
+ * But SeekableReadBuffer is not a interface which make it hard to fixup.
+ */
+class RemoteReadBuffer : public BufferWithOwnMemory<SeekableReadBufferWithSize>
+{
+public:
+    explicit RemoteReadBuffer(size_t buff_size);
+    ~RemoteReadBuffer() override;
+    static std::unique_ptr<ReadBuffer> create(ContextPtr contex, IRemoteFileMetadataPtr remote_file_metadata, std::unique_ptr<ReadBuffer> read_buffer);
+
+    bool nextImpl() override;
+    off_t seek(off_t off, int whence) override;
+    off_t getPosition() override;
+    std::optional<size_t> getTotalSize() override { return remote_file_size; }
+
+private:
+    std::shared_ptr<RemoteCacheController> file_cache_controller;
+    std::unique_ptr<ReadBufferFromFileBase> file_buffer;
+    size_t remote_file_size = 0;
+};
+
+struct RemoteFileCacheWeightFunction
+{
+    size_t operator()(const RemoteCacheController & cache) const
+    {
+        return cache.getFileSize();
+    }
+};
+
+struct RemoteFileCacheEvictPolicy
+{
+    bool canRelease(RemoteCacheController & cache) const
+    {
+        return cache.closable();
+    }
+    void release(RemoteCacheController & cache)
+    {
+        cache.close();
+    }
+};
+
+class RemoteReadBufferCache : private boost::noncopyable
```

Comment:
I think this class name is not good. It is not really about read buffer. And usually having ReadBuffer in name means deriving from some ReadBuffer* class.
May be `RemoteFilesystemCache` or `ExternalDataSourceCache` or `ExternalSourceCache` or ...
(Probably two last options better than first so not to mess with remote fs disk cache, but not necessary),

Reply:
Also I suggest to add a `Storages/Cache/` directory with separate files for this class, for cache controller class and one file for evict policy and wight function together.
So it will be like `ExternalDataSourceCache.h/cpp`, `RemoteCacheController.h/cpp`, `RemoteFileCachePolicy.h`.

Reply:
done